lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,514 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	signer := v4.NewSigner()
+	presigned, _, err := signer.PresignHTTP(ctx, creds, r, hashForGetRequests, "sts", region, timestamp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	kubeToken := awsKubePrefix + base64.RawURLEncoding.EncodeToString([]byte(presigned))
+
+	eksClusterCA, err := base64.StdEncoding.DecodeString(clusterCA)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	restConfig := &rest.Config{
+		Host:        clusterEndpoint,
+		BearerToken: kubeToken,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: eksClusterCA,
+		},
+	}
+
+	configFlags := genericclioptions.NewConfigFlags(false)
+	configFlags.WithWrapConfigFn(func(*rest.Config) *rest.Config {
+		return restConfig
+	})
+
+	return configFlags, nil
+}
+
+func getHelmActionConfig(clientGetter genericclioptions.RESTClientGetter, log logrus.FieldLogger) (*action.Configuration, error) {
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(clientGetter, agentNamespace, "secret", log.WithField("source", "helm").Debugf); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return actionConfig, nil
+}
+
+func getHelmSettings() *helmCli.EnvSettings {
+	helmSettings := helmCli.New()
+	helmSettings.RepositoryCache = os.TempDir()
+	helmSettings.SetNamespace(agentNamespace)
+
+	return helmSettings
+}
+
+// checkAgentAlreadyInstalled checks through the Helm if teleport-kube-agent chart was already installed in the EKS cluster.
+func checkAgentAlreadyInstalled(actionConfig *action.Configuration) (bool, error) {
+	listCmd := action.NewList(actionConfig)
+
+	releases, err := listCmd.Run()
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	for _, release := range releases {
+		if release.Name == agentName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// installKubeAgent installs teleport-kube-agent chart to the target EKS cluster.
+func installKubeAgent(ctx context.Context, eksCluster *eksTypes.Cluster, proxyAddr, joinToken, resourceId string, actionConfig *action.Configuration, settings *helmCli.EnvSettings, req EnrollEKSClustersRequest) error {
+	installCmd := action.NewInstall(actionConfig)
+	installCmd.RepoURL = agentRepoURL
+	installCmd.Version = req.AgentVersion
+	if strings.Contains(installCmd.Version, "dev") {
+		installCmd.Version = "" // For testing during development.
+	}
+
+	chartPath, err := installCmd.LocateChart(agentName, settings)
+	if err != nil {
+		return trace.Wrap(err, "could not locate chart")
+	}
+	defer func() {
+		// Clean up downloaded chart.
+		_ = os.Remove(chartPath)
+	}()
+
+	agentChart, err := loader.Load(chartPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	installCmd.ReleaseName = agentName
+	installCmd.Namespace = agentNamespace
+	installCmd.CreateNamespace = true
+	vals := map[string]any{}
+	vals["proxyAddr"] = proxyAddr
+
+	vals["roles"] = "kube"
+	if req.EnableAppDiscovery {
+		vals["roles"] = "kube,app,discovery"
+	}
```

Comment:
Agent Version is going to be v13 in Cloud.
Does it get the same configuration as v15?
I recall doing some checks because of this for the Kubernentes Agent's Discover Flow, and removing something about the app discovery (app role in the token?)

Maybe this is an UI thing and it shouldn't even show the EnableAppDiscovery if the target version is v13.

Reply:
Right, v13 can't handle app role without app resources defined on it. I added a check and todo about it. I don't think it's worth it to amend UI for that, as probably (hopefully 🤞 ) this problem with cloud will not persist too long and will be solved before v15 on the cloud.